### PR TITLE
Setup github actions instead of travis-ci

### PR DESCRIPTION
**Pull request by cdepillabout**
This PR gets rid of our travis-ci CI checks, and replaces them with GitHub Actions.

I've switched a couple of my other personal repositories over to GitHub Actions, and everything seems to workout pretty well.

As for documentation, other than the generic GitHub Actions documentation, the main source of documentation is the Haskell-related Action: https://github.com/actions/setup-haskell

### CODE REVIEW

**Comment by cdepillabout** [comment]
@Vlix Oh, and the GHC versions to test with were picked relatively arbitrarily, so if you'd like to remove 8.4.4 or add something like 8.2.2, please feel free to make that change.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    - name: Build
+      run: |
+        cabal update
+        cabal build all --enable-tests --enable-benchmarks --write-ghc-environment-files=always
+    - name: Test
+      run: |
+        cabal test all --enable-tests
+  stack:
+    name: stack / ghc ${{ matrix.ghc }}
```
Comment (Vlix) [inline .github/workflows/ci.yml:50]: Where does this `matrix.ghc` come from? Is that the same one from under `cabel:`?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        stack: ["latest"]
```
Comment (Vlix) [inline .github/workflows/ci.yml:54]: Would also like to `stack build` against a few LTSes and OSes like we did with travis. Not immediately sure how to pair the `--resolver lts-##.yml` with a possibly necessary `--stack-yaml stack-lts-14.yml`...

**Comment by Vlix** [comment]
Looks good, is there anywhere we can check that it works? With Travis, it'd have the handy indications above the [Merge] button of a PR like this one. Do the GitHub actions need to be initiated first? Or do you just have to go to a specific URL to see it?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (cdepillabout) [reply .github/workflows/ci.yml:50]: Ah, good catch.  It appears that there is no `matrix.ghc` for `stack`.

Let me try to fix this.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (cdepillabout) [reply .github/workflows/ci.yml:54]: I added this as well with an additional matrix entry like the following:

```yaml
        resolver:
          - "--stack-yaml ./stack.yaml"
          - "--stack-yaml ./stack-lts-14.yaml"
          - "--resolver lts-15"
          - "--resolver lts-16"
```

This says that we'll have one build using our `./stack.yaml` (just to make sure our default stack setup works), one build using our `stack-lts-14.yaml` (to make sure the build works with lts-14), one build against the latest lts-15, and one build against the latest lts-16.

(The reason for doing `--stack-yaml ./stack-lts-14.yaml` and not `--resolver lts-14` is that we need an `extra-dep` that is not in lts-14).

File: stack-lts-14.yaml [hunk 3]
```diff
@@ -17,7 +17,7 @@
 #
 # resolver: ./custom-snapshot.yaml
 # resolver: https://example.com/snapshots/2018-01-01.yaml
-resolver: lts-15.10
+resolver: lts-14.27
```
Comment (cdepillabout) [inline stack-lts-14.yaml:20]: I moved this back to lts-14.27 for the following two reasons:

- The file name is `stack-lts-14.yaml`, so I thought having this be lts-14.27 makes more sense.
- It is not possible to build `password` against lts-14 without specifying an `extra-dep`, but I don't think there is a way to do that from the command line, so it makes sense to keep around this `stack-lts-14.yaml` file, mostly for CI (assuming we do want to test on lts-14).

File: .github/workflows/ci.yml (hunk 2 above)
Comment (cdepillabout) [reply .github/workflows/ci.yml:54]: I took a shot at this in the last couple commits.

**Comment by cdepillabout** [comment]
@Vlix 

> Is there anywhere we can check that it works?

At the bottom of the screen, if you click on `Show all checks`, you should get a list of the checks that have passed:

image

After clicking:

image

It seems like the github UI doesn't necessarily do a good job of calling this out when all the tests are actually passing.

At the top of the screen there is also a `Checks` tab:

image

**Comment by cdepillabout** [comment]
@Vlix Thanks for fixing this up!

I disabled the ghc-8.2.2 build because of the persistent thing.  However, if you do really want this, let me know and we can send a PR to persistent to fix their version ranges.

Personally, I don't think it is a big deal if we don't build on ghc-8.2.2.  I think other maintainers generally try to support 3 GHC versions, so as long as we're testing on 8.6, 8.8, and 8.10, I think we're doing a pretty good job.

**Comment by cdepillabout** [comment]
Also, I'm kind of surprised that `stack build --resolver lts-11 --stack-yaml ./stack-lts-14.yaml` works correctly, but it is convenient for us that it does!

**Comment by cdepillabout** [comment]
@Vlix I think this should be ready for a final review and possibly a merge.

**Comment by Vlix** [comment]
> I disabled the ghc-8.2.2 build because of the persistent thing. However, if you do really want this, let me know and we can send a PR to persistent to fix their version ranges.

I've already created an issue asking if it's intentional that 2.11 doesn't build on 8.2, and [it isn't](https://github.com/yesodweb/persistent/issues/1180) (they also switched to GitHub Actions) so I'll probably try and add 8.2.2 to their CI at some point as well and fix whatever's breaking it. Don't see why we shouldn't add 8.2.2 afterwards, though I do agree that if any library dropped support for 8.2 and caused `password` to not build, I'd be fine with dropping 8.2 as well.

> Also, I'm kind of surprised that stack build --resolver lts-11 --stack-yaml ./stack-lts-14.yaml works correctly, but it is convenient for us that it does!

Well, the `--resolver` flag just overrides the resolver field, and the `extra-deps` stay the same, so... 👍 

---

I'll take a fresh look tomorrow and merge then.

**Comment by cdepillabout** [comment]
That sounds good to me, thanks!

**Comment by cdepillabout** [comment]
Ah, sorry, I didn't actually mean to merge this!

Well, everything was passing, so hopefully this won't negatively effect anything.

Please feel free to take a look at this and let me know if there is anything else that needs to be fixed up.  I can either just revert the merge commit, or just send another PR updating whatever needs to be updated.

**Comment by Vlix** [comment]
The only thing I can think of is maybe adding `GHC HEAD` to the `cabal` builds, but it's not too important to not merge this, so I consider this repo switched to GitHub Actions succesfully 👍

**Pull request merged**